trl/trainer/utils.py

```diff
@@ -69,43 +69,64 @@ class DataCollatorForCompletionOnlyLM(DataCollatorForLanguageModeling):
     """
 
     def __init__(
-        self,
-        response_template: str,
-        *args,
-        mlm: bool = False,
-        ignore_index: int = -100,
-        **kwargs,
+        self, instruction_template, assistant_template, *args, mlm: bool = False, ignore_index: int = -100, **kwargs
     ):
         super().__init__(*args, mlm=mlm, **kwargs)
-        self.response_template = response_template
+        self.instruction_template = instruction_template
+        self.assistant_template = assistant_template
         self.ignore_index = ignore_index
 
     def torch_call(self, examples: List[Union[List[int], Any, Dict[str, Any]]]) -> Dict[str, Any]:
         batch = super().torch_call(examples)
 
-        # The prompt ends with the response key plus a newline.  We encode this and then try to find it in the
-        # sequence of tokens.  This should just be a single token.
-        response_token_ids = self.tokenizer.encode(self.response_template, add_special_tokens=False)
+        assistant_token_ids = self.tokenizer.encode(self.assistant_template, add_special_tokens=False)
 
         labels = batch["labels"].clone()
 
         for i in range(len(examples)):
-            response_token_ids_start_idx = None
-
-            for idx in np.where(batch["labels"][i] == response_token_ids[0])[0]:
-                # `response_token_ids` is `'### Response:\n'`, here we are just making sure that the token IDs match
-                if response_token_ids == examples[i]["input_ids"][idx : idx + len(response_token_ids)]:
-                    response_token_ids_start_idx = idx
-
-            if response_token_ids_start_idx is None:
+            assistant_token_ids_idxs = []
+            human_token_ids_idxs = []
+
+            for assistant_idx in np.where(batch["labels"][i] == assistant_token_ids[0])[0]:
+                # find the indexes of the start of an assistant answer.
+                if (
+                    assistant_token_ids
+                    == examples[i]["input_ids"][assistant_idx : assistant_idx + len(assistant_token_ids)]
+                ):
+                    assistant_token_ids_idxs.append(assistant_idx + len(assistant_token_ids))
+
+            if len(assistant_token_ids) == 0:
                 raise RuntimeError(
-                    f'Could not find response key {response_token_ids} in token IDs {batch["labels"][i]}'
+                    f'Could not find response key {assistant_token_ids} in token IDs {batch["labels"][i]}'
                 )
 
-            response_token_ids_end_idx = response_token_ids_start_idx + len(response_token_ids)
+            if self.instruction_template is not None:
+                human_token_ids = self.tokenizer.encode(self.instruction_template, add_special_tokens=False)
```

Comment:
Maybe add this in `init` so that it just has to happens once and not on every call

Reply:
Yes, you can add `self.human_token_ids = self.tokenizer.encode(self.instruction_template, add_special_tokens=False)` at the init, good point.